sui_types/src/messages.rs

```diff
@@ -311,35 +313,272 @@ where
+                    let sub = kind.input_objects()?;
+                    debug_assert_eq!(sub.len(), kind.input_object_count());
+                    result.extend(sub);
+                }
+                result
+            }
+        };
+        inputs.push(InputObjectKind::OwnedMoveObject(
+            *self.gas_payment_object_ref(),
+        ));
+        Ok(inputs)
+    }
+
+    pub fn single_transactions(&self) -> impl Iterator<Item = &SingleTransactionKind> {
+        match &self.data.kind {
+            TransactionKind::Single(s) => Either::Left(std::iter::once(s)),
+            TransactionKind::Batch(b) => Either::Right(b.iter()),
+        }
+    }
+
+    pub fn into_single_transactions(self) -> impl Iterator<Item = SingleTransactionKind> {
+        match self.data.kind {
+            TransactionKind::Single(s) => Either::Left(std::iter::once(s)),
+            TransactionKind::Batch(b) => Either::Right(b.into_iter()),
+        }
+    }
+
+    // Derive a cryptographic hash of the transaction.
+    pub fn digest(&self) -> TransactionDigest {
+        TransactionDigest::new(sha3_hash(&self.data))
+    }
+
+    pub fn input_objects_in_compiled_modules(
+        compiled_modules: &[CompiledModule],
+    ) -> Vec<InputObjectKind> {
+        let mut dependent_packages = BTreeSet::new();
+        for module in compiled_modules.iter() {
+            for handle in module.module_handles.iter() {
+                let address = ObjectID::from(*module.address_identifier_at(handle.address));
+                if address != ObjectID::ZERO {
+                    dependent_packages.insert(address);
+                }
+            }
+        }
+
+        // We don't care about the digest of the dependent packages.
+        // They are all read-only on-chain and their digest never changes.
+        dependent_packages
+            .into_iter()
+            .map(InputObjectKind::MovePackage)
+            .collect::<Vec<_>>()
+    }
+}
+
+// In combination with #[serde(remote = "TransactionEnvelope")].
+// Generic types instantiated multiple times in the same tracing session requires a work around.
+// https://novifinancial.github.io/serde-reflection/serde_reflection/index.html#features-and-limitations
+impl<'de, T> Deserialize<'de> for TransactionEnvelope<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        TransactionEnvelope::deserialize(DeserializeNameAdapter::new(
+            deserializer,
+            std::any::type_name::<TransactionEnvelope<T>>(),
+        ))
+    }
+}
+
+impl<T> Serialize for TransactionEnvelope<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        TransactionEnvelope::serialize(
+            self,
+            SerializeNameAdapter::new(serializer, std::any::type_name::<TransactionEnvelope<T>>()),
```

Comment:
I suspect that instead of `std::any::type_name::<TransactionEnvelope<T>>()` what you actually want is:
```rust
format!("TransactionEnvelope<{}>", std::any::type_name::<T>().split("::").last().expect("split output should never be empty!"))
```

Look at the YAML you're generating for why.

Reply:
Unfortunately `SerializeNameAdapter` expects a static &str.

Reply:
I suspect that with the [`const_format` crate], you can still do something like:
```rust
use const_format::formatcp;
...
 &formatcp!("TransactionEnvelope<{}>", std::any::type_name::<T>())
 ```
 
 Which would be a huge usability improvement.

Reply:
hmm, this doesn't work:
```
            &formatcp!("TransactionEnvelope<{}>", std::any::type_name::<T>()),

```
gives
```
type annotations needed
type must be known at this pointrustc[E0282](https://doc.rust-lang.org/error-index.html#E0282)
[helper_macros.rs(80, 17): ]()Actual error occurred here
[fmt_macros.rs(87, 41): ]()Error originated from macro call here
[fmt_macros.rs(262, 9): ]()Error originated from macro call here
[fmt_macros.rs(262, 9): ]()Error originated from macro call here
[fmt_macros.rs(76, 17): ]()consider giving `out` a type
```

Reply:
@huitseeker can't we use that name_variant module you came up with for this?

Reply:
Yes, we could but at this stage, I'd rather punt and open an issue to mop this up. 😓